File: flink-demo/flinkScalaCode/src/main/scala/com/yw/flink/example/scalacases/case04_sink/Case04_RedisSink.scala
package com.yw.flink.example.scalacases.case04_sink

import org.apache.flink.streaming.api.scala.StreamExecutionEnvironment
import org.apache.flink.streaming.connectors.redis.RedisSink
import org.apache.flink.streaming.connectors.redis.common.config.FlinkJedisPoolConfig
import org.apache.flink.streaming.connectors.redis.common.mapper.{RedisCommand, RedisCommandDescription, RedisMapper}

/**
  * Flink写出数据到Redis
  */
object Case04_RedisSink {
  def main(args: Array[String]): Unit = {
    val env: StreamExecutionEnvironment = StreamExecutionEnvironment.getExecutionEnvironment
    import org.apache.flink.streaming.api.scala._
    val result: DataStream[(String, Int)] = env.socketTextStream("node5", 9999)
      .flatMap(_.split(","))
      .map((_, 1))
      .keyBy(_._1)
      .sum(1)

    //准备连接Redis配置
    val conf: FlinkJedisPoolConfig = new FlinkJedisPoolConfig.Builder()
      .setHost("node4")
      .setPort(6379)
      .setDatabase(2)
      .build()

    //准备RedisSink
    val redisSink = new RedisSink[(String, Int)](conf, new RedisMapper[(String, Int)] {
      //指定Redis的结构和key
      override def getCommandDescription: RedisCommandDescription = {
        new RedisCommandDescription(RedisCommand.HSET, "flink-scala-redis")
      }

      //指定redis hset 中的key
      override def getKeyFromData(t: (String, Int)): String = t._1

      //指定redis hset中的value
      override def getValueFromData(t: (String, Int)): String = t._2.toString

    })

    result.addSink(redisSink)
    env.execute()
  }
}

File: flink-demo/flinkJavaCode/src/main/java/com/yw/flink/example/javacases/case04_sink/Case04_RedisSink.java
package com.yw.flink.example.javacases.case04_sink;

import org.apache.flink.api.common.typeinfo.Types;
import org.apache.flink.api.java.tuple.Tuple2;
import org.apache.flink.streaming.api.datastream.SingleOutputStreamOperator;
import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;
import org.apache.flink.streaming.connectors.redis.RedisSink;
import org.apache.flink.streaming.connectors.redis.common.config.FlinkJedisPoolConfig;
import org.apache.flink.streaming.connectors.redis.common.mapper.RedisCommand;
import org.apache.flink.streaming.connectors.redis.common.mapper.RedisCommandDescription;
import org.apache.flink.streaming.connectors.redis.common.mapper.RedisMapper;
import org.apache.flink.util.Collector;

/**
 * Flink将数据写出到Redis
 */
public class Case04_RedisSink {
    public static void main(String[] args) throws Exception {
        StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();
        SingleOutputStreamOperator<Tuple2<String, Integer>> result = env.socketTextStream("nc_server", 9999)
                .flatMap((String line, Collector<String> collector) -> {
                    String[] split = line.split(",");
                    for (String word : split) {
                        collector.collect(word);
                    }

                }).returns(Types.STRING)
                .map(word -> Tuple2.of(word, 1)).returns(Types.TUPLE(Types.STRING, Types.INT))
                .keyBy(tp -> tp.f0)
                .sum(1);

        //连接Redis的配置
        FlinkJedisPoolConfig conf = new FlinkJedisPoolConfig.Builder()
                .setHost("node4")
                .setPort(6379)
                .setDatabase(1)
                .build();
        //准备RedisSink

        RedisSink<Tuple2<String, Integer>> redisSink = new RedisSink<>(conf, new RedisMapper<Tuple2<String, Integer>>() {
            //指定写出到Redis什么结构、什么表
            @Override
            public RedisCommandDescription getCommandDescription() {
                return new RedisCommandDescription(RedisCommand.HSET, "flink-redis-sink");
            }

            //指定Redis写出的key
            @Override
            public String getKeyFromData(Tuple2<String, Integer> tp) {
                return tp.f0;
            }

            //指定Redis写出的value
            @Override
            public String getValueFromData(Tuple2<String, Integer> tp) {
                return tp.f1 + "";
            }
        });

        result.addSink(redisSink);

        env.execute();

    }
}
